audio-device-client/explainer.md

```diff
@@ -0,0 +1,167 @@
+newly added Audio Worklet, but it is still confined by the boundary of Web Audio
+API's graph rendering mechanism.
+
+To overcome this limitation, some "pro-audio" web apps were designed with a new
+processing model and it bypasses the most of Web Audio API's graph system by
+utilizing Audio Worklet, SharedArrayBuffer and Worker.
+
+![Design pattern: WebAudio Powerhouse](https://github.com/WebAudio/web-audio-cg/blob/master/audio-device-client/images/img-1-design-pattern.png "Design pattern: WebAudio Powerhouse")
+
+This setup is convenient to take advantage of the code compiled into
+WebAssembly, therefore this new model cuts the engineering cost significantly
+because audio developers can bring existing source codes to the web platform
+with minimum effort. Additionally, deploying to multiple targets by using same
+source code ensures the identical sonic result across various platforms.
+(e.g. encoding/decoding, signal processing or an entire audio application)
+
+However, this convoluted workaround only solves some of problems that this breed
+of audio-centric applications faces. It is still locked by the graph render
+quantum size (128 sample-frames), and there is no unified API that configures
+essential properties of audio rendering mechanism (such as I/O device selection,
+multi-channel I/O support, configurable sample rate and more).
+
+The slow adoption of Web Audio API from pro-audio or game industry is the clear
+evidence of why this problem is important to solve.
+
+
+## Key Features
+
+- Provides a dedicated global scope on a separate thread
+  (real-time/high-priority when permitted) for audio processing purpose
+- Selecting audio I/O devices can be done via `MediaTrackConstraints` pattern
+- Supports variable callback buffer size (as opposed to 128 sample-frames
+  limitation of Web Audio API)
+- Provides implicit sample rate conversion when needed
+- Serves I/O audio data from hardware in a single callback function
+- Provides an AudioContext instance when requested
+- HTTPS only and follows Autoplay policy
+
+
+## Non-goals
+
+- Is NOT to replace Web Audio API
+- Contains no media encoding/decoding service
+
+
+## Potential Use Cases
+
+- A __teleconference app__ with custom audio processing and direct hardware
+  access (e.g. complex echo cancellation, source separation, audio
+  spatialization, and auditory scene analysis/music information retrieval)
+- Porting __existing audio engines__ to the web platform with the minimum
+  engineering cost.
+- A __hybrid audio application__ that uses Web Audio API's graph system
+  (processing and synthesis) and customized audio hardware configuration.
+
+
+## How Audio Device Client Works
+
+![Audio Device Client](https://github.com/WebAudio/web-audio-cg/blob/master/audio-device-client/images/img-2-audio-device-client.png "Audio Device Client")
+
+As discussed, the Audio Device Client will be an intermediate layer between Web
```

Comment:
Is this really true? (intermediate layer between WebAudio and audio devices)  I thought we really wanted that to be completely separate.

Reply:
This is something discussed in last TPAC. But if you want to revisit this, we can open an issue and talk about it.